### ISSUE TITLE
Adding password to dsn causes scripts to ignore CTRL-C

### ISSUE THREAD

**Issue by richardbann** [psycopg/psycopg2#898]
OS version: Ubuntu 18.04
psycopg2 version: 2.8.1

The following script can be stopped with CTRL-C:

```py
import psycopg2


while True:
    try:
        psycopg2.connect("host=postgres")
    except Exception as e:
        continue
    else:
        break
```

Adding a password causes the script to ignore CTRL-C:

```py
import psycopg2


while True:
    try:
        psycopg2.connect("host=postgres password=x")
    except Exception as e:
        continue
    else:
        break
```

Please help, what makes such a big difference?

**Comment by dvarrazzo** [comment]
Funny, it's true.

The command in itself terminates on error (if the password is wrong):

```
In [4]: psycopg2.connect("host=localhost user=wat dbname=piro password=y")
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
<ipython-input-4-f5b524a63be1> in <module>()
----> 1 psycopg2.connect("host=localhost user=wat dbname=piro password=y")

/home/piro/dev/psycopg2/build/lib.2.7/psycopg2/__init__.pyc in connect(dsn, connection_factory, cursor_factory, **kwargs)
    124 
    125     dsn = _ext.make_dsn(dsn, **kwargs)
--> 126     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
    127     if cursor_factory is not None:
    128         conn.cursor_factory = cursor_factory

OperationalError: FATAL:  password authentication failed for user "wat"
FATAL:  password authentication failed for user "wat"
```

But in a loop it doesn't receive a SIGINT:

```
In [5]: while True:
   ...:     try:
   ...:         psycopg2.connect("host=localhost user=wat dbname=piro password=y")
   ...:     except Exception as e:
   ...:         continue
   ...:     else:
   ...:         break
   ...:     
^C^C^C^CTerminated
```

(until killed with a TERM).

If I add a couple of prints, before the `connect()` and after the `except`, I can see the command in a tight loop, and pressing ctrl-C the output is "disturbed" by the `^C` chars appearing and streaming up. So it doesn't seem a psycopg bug, rather a Python one. Maybe the quick release and acquisition of the GIL doesn't allow the signal handler to operate correctly.

Anyway, if you put a `time.sleep(0.01)` after the `except` it will receive and handle the ctrl-c ok.

**Comment by richardbann** [comment]
Thanks for looking into it.

It is very unlikely it is a python bug as it works as expected with psycopg2 2.7.1 and the described behavior happens starting with 2.7.2.

**Comment by richardbann** [comment]
After some testing it seems to be this commit: 9f160fd

**Comment by dvarrazzo** [comment]
Interesting: probably catching and re-rising the exception does something.

The problem caused is really marginal though: there isn't really a reason to loop so tightly around trying to reconnect. If the code is correct (e.g. it doesn't swallow an exception) I'm not interested to investigate further.

**Comment by dvarrazzo** [comment]
As discussed in #934, the issue is not related to a tight loop but to the exception swallowing, in order to obfuscate the password but ignore eventual errors in that function in order to avoid clobbering connection errors, deemed more important.

Issue fixed by simply obfuscating the password before attempting the connection, and not tampering with the exception altogether. To be released in 2.8.4.
